File: 1008. Construct Binary Search Tree from Preorder Traversal/tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* bstFromPreorder(vector<int>& preorder) {
        return build(preorder, 0, preorder.size() - 1);
    }
    TreeNode* build(const vector<int>& preorder, int low, int high){
        //cout << "low:" << low << ", high:" << high << endl;
        if(low > high)
            return nullptr;
        if(low == high)
            return new TreeNode(preorder[low]);
        int rootVal = preorder[low];
        int pos = findPos(preorder, low, high, rootVal);
        //cout << "find:" << pos << endl;
        TreeNode *root = new TreeNode(rootVal);
        root->left = build(preorder, low + 1, pos - 1);
        root->right = build(preorder, pos, high);
        return root;
    }
    int findPos(const vector<int>& preorder, int low, int high, int target){
        int i = low;
        for(; i <= high; ++i){
            if(preorder[i] > target)
                break;
        }
        return i;
    }
};

File: offer/JZ33 二叉搜索树的后序遍历序列/tree.java
public class Solution {
    private int findIndex(int[] v, int low, int high, int target){
        for(int i = high - 1; i >= low; i--){
            if(v[i] < target)
                return i;
        }
        return low - 1;
    }
    public boolean verify(int[] v, int low, int high){
        //System.out.println(low + "," + high);
        if(low > high)
            return true;
        if(low == high)
            return true;
        int rootVal = v[high];
        int index = findIndex(v, low, high, rootVal);
        // left node is null
        if(index == low - 1)
            return verify(v, low, high - 1);
        if(index == high - 1){
            // right node is null
            for(int i = low; i <= high - 1; ++i){
                if(v[i] > rootVal)
                    return false;
            }
            return verify(v, low, high - 1);
        }
        // left node check
        for(int i = low; i <= index; ++i){
            if(v[i] > rootVal)
                return false;
        }
        // right node check
        for(int i = index + 1; i <= high - 1; ++i){
            if(v[i] < rootVal)
                return false;
        }
        return verify(v, low, index) && verify(v, index + 1, high - 1);
        
    }
    public boolean VerifySquenceOfBST(int [] v) {
        if(v.length == 0)
            return false;
        return verify(v, 0, v.length - 1);
    }
}